File: CONTAIN/Ball.m
//
//  Ball.m
//  CONTAIN
//
//  Created by Quinton Ashley on 1/13/15.
//  Copyright (c) 2015 Quinton Ashley. All rights reserved.
//

#import "Ball.h"

@implementation Ball

+ (id)newBallWithRadius:(CGFloat)radius position:(CGPoint)position speed:(CGFloat)speed {
  Ball *ball = [Ball spriteNodeWithColor:[SKColor whiteColor] size:CGSizeMake(radius, radius)];
  ball.physicsBody = [SKPhysicsBody bodyWithCircleOfRadius:radius];
  float velX = arc4random_uniform(speed/2);
  float velY = speed-velX;
  int quadrantChance = arc4random_uniform(100);
  if (quadrantChance <= 25) {
    ball.physicsBody.velocity = CGVectorMake(velX, velY);
  } else if (quadrantChance <= 50) {
    ball.physicsBody.velocity = CGVectorMake(-velX, velY);
  }  else if (quadrantChance <= 75) {
    ball.physicsBody.velocity = CGVectorMake(-velX, -velY);
  }  else {
    ball.physicsBody.velocity = CGVectorMake(velX, -velY);
  }
  ball.physicsBody.friction = 0;
  ball.physicsBody.restitution = 1;
  ball.physicsBody.linearDamping = 0;
  ball.position = position;
  ball.physicsBody.categoryBitMask = 0x1 << 0;
  ball.physicsBody.collisionBitMask = 0x1 << 10;
  ball.physicsBody.contactTestBitMask = 0x1 << 1 | 0x1 << 2;
  return ball;
}

+ (id)newBallWithRadius:(CGFloat)radius position:(CGPoint)position vector:(CGVector)vector {
  Ball *ball = [Ball spriteNodeWithColor:[SKColor whiteColor] size:CGSizeMake(radius, radius)];
  ball.physicsBody = [SKPhysicsBody bodyWithCircleOfRadius:radius];
  ball.physicsBody.velocity = vector;
  ball.physicsBody.friction = 0;
  ball.physicsBody.restitution = 1;
  ball.physicsBody.linearDamping = 0;
  ball.position = position;
  ball.physicsBody.categoryBitMask = 0x1 << 0;
  ball.physicsBody.collisionBitMask = 0x1 << 10;
  ball.physicsBody.contactTestBitMask = 0x1 << 1 | 0x1 << 2;
  return ball;
}

@end

File: Ball.swift
//  Converted to Swift 5.1 by Swiftify v5.1.30744 - https://objectivec2swift.com/
//
//  Ball.m
//  CONTAIN
//
//  Created by Quinton Ashley on 1/13/15.
//  Copyright (c) 2015 Quinton Ashley. All rights reserved.
//
import SpriteKit

class Ball: SKSpriteNode {
	init(radius: CGFloat, speed: CGFloat) {
		super.init(texture: SKTexture(imageNamed: "ball"), color: SKColor.clear, size: CGSize(width: radius*2, height: radius*2));
		alpha = 1;
		physicsBody = SKPhysicsBody(circleOfRadius: radius);
		let velX = CGFloat(arc4random_uniform(UInt32(speed / 2)));
		let velY = CGFloat(speed - CGFloat(velX));
		let quadrantChance = arc4random_uniform(100)
		if (quadrantChance <= 25) {
			physicsBody?.velocity = CGVector(dx: velX, dy: velY);
		} else if (quadrantChance <= 50) {
			physicsBody?.velocity = CGVector(dx: -velX, dy: CGFloat(velY));
		} else if (quadrantChance <= 75) {
			physicsBody?.velocity = CGVector(dx: -velX, dy: -velY);
		} else {
			physicsBody?.velocity = CGVector(dx: velX, dy: -velY);
		}

		physicsBody?.friction = 0;
		physicsBody?.restitution = 1;
		physicsBody?.linearDamping = 0;
		physicsBody?.categoryBitMask = 0x1 << 0;
		physicsBody?.collisionBitMask = 0x1 << 10;
		physicsBody?.contactTestBitMask = 0x1 << 1 | 0x1 << 2;
		physicsBody?.usesPreciseCollisionDetection = true;
	}
	
	init(radius: CGFloat, vector: CGVector) {
		super.init(texture: SKTexture(imageNamed: "ball"), color: SKColor.clear, size: CGSize(width: radius*2, height: radius*2));
		physicsBody = SKPhysicsBody(circleOfRadius: radius);
		physicsBody?.velocity = vector;
		physicsBody?.friction = 0;
		physicsBody?.restitution = 1;
		physicsBody?.linearDamping = 0;
		physicsBody?.categoryBitMask = 0x1 << 0;
		physicsBody?.collisionBitMask = 0x1 << 10;
		physicsBody?.contactTestBitMask = 0x1 << 1 | 0x1 << 2;
	}
	
  required public init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
}
